### PR TITLE
feat(sweeps): Sweep scheduler cleanup and better tests

### PR DESCRIPTION
**Pull request by hu-po**
Fixes WB-10588

Description
-----------
Cleans up some of the threading logic in the sweep scheduler. Adds better unit tests for sweep scheduler.

Testing
-------
How was this PR tested?

Checklist
-------
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/launch/sweeps/scheduler.py [hunk 1]
```diff
@@ -195,11 +192,16 @@ def _update_run_states(self) -> None:
                 ]:
                     run.state = SimpleRunState.ALIVE
             except Exception as e:
```
Comment (gtarpenning) [inline wandb/sdk/launch/sweeps/scheduler.py:194]: If this happens relatively frequently you could try to catch the specific Exception here, especially if its a network/comm thing that happens non-deterministically.

File: wandb/sdk/launch/sweeps/scheduler.py [hunk 2]
```diff
@@ -208,28 +210,27 @@ def _add_to_launch_queue(
     ) -> "public.QueuedRun":
         """Add a launch job to the Launch RunQueue."""
         run_id = run_id or generate_id()
+        # One of Job and URI is required
+        _job = self._kwargs.get("job", None)
+        _uri = self._kwargs.get("uri", None)
+        if _job is None and _uri is None:
+            # If no Job is specified, use a placeholder URI to prevent Launch failure
```
Comment (gtarpenning) [inline wandb/sdk/launch/sweeps/scheduler.py:217]: @KyleGoyette I wonder if there is a better way of handling this on the Launch side, I also have had to write temp URI's which kinda defeats the purpose of the check.

File: wandb/sdk/launch/sweeps/scheduler_sweep.py [hunk 3]
```diff
@@ -38,120 +37,114 @@ class SweepScheduler(Scheduler):
...
+            {},  # metrics: dict
+            _run_states,  # run_states: dict
+        )
+        logger.debug(
+            f"{LOG_PREFIX}AgentHeartbeat received {len(commands)} commands: \n{pprint.pformat(commands)}\n"
+        )
+        if commands:
+            for command in commands:
+                # The command "type" can be one of "run", "resume", "stop", "exit"
+                _type = command.get("type", None)
```
Comment (gtarpenning) [inline wandb/sdk/launch/sweeps/scheduler_sweep.py:98]: Is this better style than just `command.get("type")` ? Purely curious

Comment (hu-po) [reply wandb/sdk/launch/sweeps/scheduler_sweep.py:98]: Probably a wash. It is more readable/obvious at the cost of extra verbosity.

File: tests/unit_tests/test_sweep_scheduler.py [hunk 4]
```diff
@@ -21,114 +24,153 @@ def test_sweep_scheduler_load():
...
-        "wandb.sdk.launch.sweeps.scheduler.Scheduler._run",
-        mock_run_raise_keyboard_interupt,
-    )
+@patch.multiple(Scheduler, __abstractmethods__=set())
+@pytest.mark.parametrize("sweep_config", VALID_SWEEP_CONFIGS_MINIMAL)
+def test_sweep_scheduler_base_scheduler_states(
+    user, relay_server, sweep_config, monkeypatch
+):
+
+    with relay_server():
```
Comment (KyleGoyette) [inline tests/unit_tests/test_sweep_scheduler.py:49]: Is this the new server??!

File: wandb/sdk/launch/sweeps/scheduler.py (hunk 1 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/sweeps/scheduler.py:194]: I think the only time this happens is when the run doesn't exist yet right? Still we should attempt to be as narrow as possible with our catch exceptions.

File: wandb/sdk/launch/sweeps/scheduler.py (hunk 2 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/sweeps/scheduler.py:217]: Yeah, I guess this pattern is used when we just want to run command in a process that doesn't refer to a run or file? This is fine for now, but we should potentially think about that. Maybe just check if there's an entrypoint?

File: wandb/sdk/launch/sweeps/scheduler_sweep.py [hunk 5]
```diff
@@ -38,120 +37,114 @@ class SweepScheduler(Scheduler):
     def __init__(
         self,
         *args: Any,
-        sweep_id: Optional[str] = None,
         num_workers: int = 4,
-        heartbeat_thread_sleep: float = 0.5,
-        heartbeat_queue_timeout: float = 1,
-        main_thread_sleep: float = 1,
+        worker_sleep: float = 0.1,
```
Comment (KyleGoyette) [inline wandb/sdk/launch/sweeps/scheduler_sweep.py:41]: These are very short sweeps, just so I can be educated, why are we reducing them so?

**Review by KyleGoyette — APPROVED** [review]
LGTM, but there are some questions that I think warrant discussion

File: tests/unit_tests/test_sweep_scheduler.py (hunk 4 above)
Comment (hu-po) [reply tests/unit_tests/test_sweep_scheduler.py:49]: mmmmhmmm

File: wandb/sdk/launch/sweeps/scheduler_sweep.py (hunk 5 above)
Comment (hu-po) [reply wandb/sdk/launch/sweeps/scheduler_sweep.py:41]: Good catch, I had them short for testing purposes. 1 seconds seems like a better default.

**Pull request merged**